kexgssc.c

```diff
@@ -453,11 +453,15 @@ kexgssgex_client(struct ssh *ssh)
 			do {
 				type = ssh_packet_read(ssh);
 				if (type == SSH2_MSG_KEXGSS_HOSTKEY) {
+					char *tmp = NULL;
+					size_t tmp_len = 0;
 					debug("Received KEXGSS_HOSTKEY");
 					if (server_host_key_blob)
 						fatal("Server host key received more than once");
-					if ((r = sshpkt_getb_froms(ssh, &server_host_key_blob)) != 0)
+					if ((r = sshpkt_get_string(ssh, &tmp, &tmp_len)) != 0)
 						fatal("sshpkt failed: %s", ssh_err(r));
+					if ((server_host_key_blob = sshbuf_from(tmp, tmp_len)) == NULL)
+						fatal("sshbuf_from failed");
 				}
 			} while (type == SSH2_MSG_KEXGSS_HOSTKEY);
 
```

Comment:
Shouldn't that change be made in the `kexgss_client()` function? GSS GEX was always working for me and actually the client also always worked for me.

Reply:
Because the servers are not sending hostkeys as part of the gss key exchange:

https://github.com/openssh-gsskex/openssh-gsskex/pull/19/files#diff-67c1dce8b9dfa4db6593655232d7fdcffd2a607e9ebd3522a7d2cf97e1d26547R137

I am not sure if there is actually any other implementation that would do that.

Reply:
> Because the servers are not sending hostkeys as part of the gss key exchange:
> 
> https://github.com/openssh-gsskex/openssh-gsskex/pull/19/files#diff-67c1dce8b9dfa4db6593655232d7fdcffd2a607e9ebd3522a7d2cf97e1d26547R137

Interesting. But when I had packet debugging enabled I could see the server sending packets that included data that looked like a certificate or a certificate chain as it included the server's DN and the signer's DN and the DN that signed the signer. Strange. :-/

But still, the change is in the wrong function - `kexgssgex_client()` instead of `kexgss_client()` - , or do I misunderstand something here?

Apart from that I just built a GSI-OpenSSH 8.0p1 with the patch to `kexgss_server()` and...

....this works! :-)

```
[johndoe@host ~]$ gsissh -vvv -oGSSAPIKexAlgorithms=gss-curve25519-sha256- -p 2222 host.domain.tld
OpenSSH_8.0p1c-GSI GSI-hpn14v19, OpenSSL 1.1.1c FIPS  28 May 2019
[...]
debug1: Local version string SSH-2.0-OpenSSH_8.0p1c-GSI GSI-hpn14v19
debug1: Remote protocol version 2.0, remote software version OpenSSH_8.0p1c-GSI GSI-hpn14v19
[...]
debug1: kex: algorithm: gss-curve25519-sha256-vz8J1E9PzLr8b1K+0remTg==
debug1: kex: host key algorithm: ecdsa-sha2-nistp256
debug1: REQUESTED ENC.NAME is 'aes256-gcm@openssh.com'
debug1: kex: server->client cipher: aes256-gcm@openssh.com MAC: <implicit> compression: none
debug1: REQUESTED ENC.NAME is 'aes256-gcm@openssh.com'
debug1: kex: client->server cipher: aes256-gcm@openssh.com MAC: <implicit> compression: none
debug1: kex: gss-curve25519-sha256-vz8J1E9PzLr8b1K+0remTg== need=32 dh_need=32
debug1: kex: gss-curve25519-sha256-vz8J1E9PzLr8b1K+0remTg== need=32 dh_need=32
debug1: Calling gss_init_sec_context
debug1: Delegating credentials
debug3: send packet: type 30
debug3: receive packet: type 31
debug1: Received GSSAPI_CONTINUE
debug1: Calling gss_init_sec_context
debug1: Delegating credentials
debug3: send packet: type 31
debug3: receive packet: type 31
debug1: Received GSSAPI_CONTINUE
debug1: Calling gss_init_sec_context
debug1: Delegating credentials
debug3: send packet: type 31
debug3: receive packet: type 31
debug1: Received GSSAPI_CONTINUE
debug1: Calling gss_init_sec_context
debug1: Delegating credentials
debug3: send packet: type 31
debug3: receive packet: type 32
debug1: Received GSSAPI_COMPLETE
[...]
debug1: Authentication succeeded (gssapi-keyex).
Authenticated to host.domain.tld ([172.16.1.34]:2222).
[...]
[johndoe2@host ~]$ 
```

Reply:
Oh, right. It should be in both of them as both of them use the `sshpkt_getb_froms()` function which is causing this issue. I was putting this together late in the evening yesterday so I did not notice it.

Reply:
Ok, I created new GSI-OpenSSH RPMs with these client changes in and the resulting binaries also work for me (tested both with `gss-curve25519-sha256-` and `gss-gex-sha1-` methods). Great work @Jakuje! :-)

How should we integrate that into GSI-OpenSSH 8.0p1 (for EPEL8)? Should I provide the changes from this PR here as additional patch to the maintainer (@ellert) or should we just adapt the "openssh-8.0p1-gssapi-keyex.patch" patch over at https://src.fedoraproject.org/rpms/gsi-openssh/tree/epel8?

I will also go through the newer GSI-OpenSSH versions - 8.3p1 and 8.4p1 ATM - and check with the changes from this PR here.

****

Slightly off topic: Is GSS group exchange generally flawed or why wasn't a SHA256 based version implemented?